### ISSUE TITLE
PluginBrowserViewModel doesn't work if the activity is destroyed

### ISSUE THREAD

**Issue by oguzkocer** [wordpress-mobile/WordPress-Android#7237]
The thing about ViewModel is that it survives the configuration changes, however if the activity is destroyed it will not survive, so we lose all the data we'd need to recreate the view. This [blog post](https://proandroiddev.com/customizing-the-new-viewmodel-cf28b8a7c5fc) describes the issue better than I could and proposes a way to fix it. I do not want to include this in #7229, so I am opening a separate issue which should be handled after @malinajirka's PR #7232 is merged which prepares us for the solution.

**Comment by oguzkocer** [comment]
@malinajirka I've been working on this issue and so far have it working with https://github.com/wordpress-mobile/WordPress-Android/commit/d41afa385f219bc7afb86d91753f5339b1eed611, do you think there might be a better solution (similar to that blog post) we can use with dependency injection? I am curious if there is a way to pass the `savedInstanceState` of the activity to the ViewModel's constructor somehow. I guess we could use our own Bundle instance and pass it during construction, but I don't know how much I like that approach or where that bundle would be saved.

If we can't do any of that ^, I am thinking about moving some of this solution to an interface. We could potentially use a base class, but those are evil and I'd much prefer to stay away from them if we can.

**Comment by oguzkocer** [comment]
I have been thinking a lot about this and if we can pass the `Intent` and the `savedInstanceState` to the constructor of the ViewModel, we can have a pretty clean architecture. I'd love to get rid of the `start` method in the model (although it's recommended) and handle everything in the constructor.

**Comment by malinajirka** [comment]
I've been thinking about it. There are few questions that comes to my mind.
>I have been thinking a lot about this and if we can pass the Intent and the savedInstanceState to the constructor of the ViewModel

ViewModel outlives it's parent activity in a retained fragment -> if the scope of the viewModel would be the same as the scope of the activity, we could create a new dagger component with a custom @PerActivity scope. We could pass the Intent and the savedInstanceState into the constructor of the component -> and inject it into the ViewModel. Unfortunately this is not the case, as the 
ViewModel can have multiple parent activities during it's lifetime :(. 

Do we really need to save the state of the ViewModel into a 'semi-persistent storage(bundle)'? Are there any benefits? I've always followed this pattern - https://developer.android.com/topic/libraries/architecture/saving-states.html - in my small testing projects.

Other semi-related thing is, that I'm still not sure of whether it's a good idea to have Android Framework dependencies in ViewModels. I've always tried to keep ViewModel's framework independent as much as possible - mostly due to testability. I'm not sure if having Bundle and Intent in the ViewModel will cause any real troubles in tests. WDYT?

**Comment by oguzkocer** [comment]
> Do we really need to save the state of the ViewModel into a 'semi-persistent storage(bundle)'? Are there any benefits?

Yeap, as mentioned in the issue description, if the activity is destroyed, we'll lose all the data. We need a few key pieces to survive to be able to recreate the activity.

> I've always followed this pattern - https://developer.android.com/topic/libraries/architecture/saving-states.html - in my small testing projects.

That last section `Restoring complex states: reassembling the pieces` in that guide actually handles the issue :)

> I'm not sure if having Bundle and Intent in the ViewModel will cause any real troubles in tests. WDYT?

If we injected it, it wouldn't be an issue, because we could inject whatever we needed for tests.

For now, I think we should go with the solution in `plugin-browser-viewmodel-bundle` and refactor it when we find a better way. We shouldn't hold the feature to find the perfect solution to the problem. We can keep the discussion going on in this thread or elsewhere though to keep looking for a better way.

**Comment by malinajirka** [comment]
>Do we really need to save the state of the ViewModel into a 'semi-persistent storage(bundle)'? Are there any benefits?

>>Yeap, as mentioned in the issue description, if the activity is destroyed, we'll lose all the data. We need a few key pieces to survive to be able to recreate the activity.

Yeah, but as far as I understand google's solution, if the activity is destroyed due to low memory for example, they save the query in the activity and pass it to the ViewModel. So they don't try to save all the loaded data into the bundle, they save just the query. When a new instance of the ViewModel is created, they pass the query to it and the ViewModels fetches the corresponding data from the repository (db/network/file/etc).
Saving loaded data into the bundle is a bit risky, as the bundle has limited capacity and if there are multiple activities on the back stack it can lead to crash. 

I agree, that this solution is absolutely fine and we can merge it. I just wanted to discuss the options, as saving and restoring ViewModel's state is a lot of work in a long term run.

**Comment by oguzkocer** [comment]
> So they don't try to save all the loaded data into the bundle, they save just the query.

This is exactly what we do as well ;) The only exception being the site model (which is saved as a model everywhere) and has been a source of a lot of discussion. I suspect we'll stop doing that some time soon.

**Comment by malinajirka** [comment]
Perfect, thanks for clearing out this for me and thank you for this PR ;).
